components/sup/src/manager/service/hook_runner.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2019 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Runs a service lifecycle hook on a separate thread, and wraps the
+//! whole execution in a future.
+//!
+//! Ideally, we'd want to use something like
+//! [tokio_process](https://github.com/alexcrichton/tokio-process),
+//! but we're not able to use that based on how our Windows hooks get
+//! executed. If that were to be re-cast in terms of Rust's
+//! `std::process::Command`, we could consider it. In the meantime,
+//! this seems to do the trick.
+
+use super::{hook_timer, spawned_future::SpawnedFuture, Pkg};
+use crate::error::SupError;
+use futures::{sync::oneshot, IntoFuture};
+use habitat_common::templating::hooks::Hook;
+use habitat_core::service::ServiceGroup;
+use std::{
+    sync::{Arc, Mutex},
+    thread,
+};
+
+pub struct HookRunner<H: Hook> {
+    hook: Arc<Mutex<H>>,
+    sg: ServiceGroup,
+    pkg: Pkg,
+    passwd: Option<String>,
+}
+
+impl<H: Hook> HookRunner<H> {
+    pub fn new(
+        hook: Arc<Mutex<H>>,
+        sg: ServiceGroup,
+        pkg: Pkg,
+        passwd: Option<String>,
+    ) -> HookRunner<H> {
+        HookRunner {
+            hook: hook,
+            sg: sg,
+            pkg: pkg,
+            passwd: passwd,
+        }
+    }
+}
+impl<H: Hook + 'static> IntoFuture for HookRunner<H> {
+    type Item = H::ExitValue;
+    type Error = SupError;
+    type Future = SpawnedFuture<Self::Item>;
+
+    fn into_future(self) -> Self::Future {
+        let (tx, rx) = oneshot::channel();
+
+        // TODO (CM): Consider using a short abbreviation for the hook
+        // name in the thread name (e.g. "HC" for "health_check", "I"
+        // for "init", etc.
```

Comment:
I think `H::file_name()` seems fine and less abstruse.

Reply:
I was thinking specifically of making it a bit easier to see what's going on with each thread (on a Linux box, at least). Since most (all?) tooling seems to truncate thread names at 16 characters, this would allow you to see more of the variable parts of the thread name.

Reply:
I see. Even so, with our current hook names, we'd get between 3 and 12 characters of the service group name:
```
file-updated-ser|
health-check-ser|
init-service_gro|
run-service_grou|
post-run-service|
reload-service_g|
reconfigure-serv|
suitability-serv|
post-stop-servic|
```
If it becomes a problem in practice, we can always come back and add a short name function to `Hook`, but my guess is we won't need it.